src/core/room-lifecycle-manager.ts

```diff
@@ -1,923 +1,392 @@
-        // If in doing the wind down, we've entered failed state, then it's game over anyway
-        // TODO: Another PR, but in the even if we get a failed channel, we still need to do the wind down
-        // of other channels for atomicity.
-        // https://github.com/ably/ably-chat-js/issues/416
-        if (this._lifecycle.status === RoomStatus.Failed) {
-          throw new Error('room is in a failed state');
-        }
-
-        // If not, we wait a short period and then try again
-        return new Promise<unknown>((resolve) => {
-          setTimeout(() => {
-            resolve(this._doChannelWindDown(contributor));
-          }, 250);
+        this._logger.warn('discontinuity detected', {
+          roomId: this._roomId,
+          error,
         });
-      });
-    } catch {
-      // If an error gets through here, then the room has entered the failed state, we're done.
-      this._endLifecycleOperation();
-      return;
-    }
+        this._eventEmitter.emit(RoomEvents.Discontinuity, error);
+      }
+    });
 
-    // If our problem channel has reattached, then we can retry the attach
-    if (contributor.channel.state === 'attached') {
-      this._logger.debug('RoomLifecycleManager._doRetry(); feature reattached, retrying attach');
-      return doAttachWithRetry();
-    }
+    // CHA-RL12a, CHA-RL12b
+    channel.on('update', (stateChange: Ably.ChannelStateChange) => {
+      if (
+        !stateChange.resumed &&
+        this._hasAttachedOnce &&
+        !this._isExplicitlyDetached &&
+        stateChange.current === 'attached' &&
+        stateChange.previous === 'attached'
+      ) {
+        const error = new Ably.ErrorInfo(
+          'discontinuity detected',
+          ErrorCodes.RoomDiscontinuity,
+          stateChange.reason?.statusCode ?? 0,
+          stateChange.reason,
+        );
 
-    // Otherwise, wait for our problem channel to re-attach and try again
-    return new Promise<void>((resolve) => {
-      const listener = (change: Ably.ChannelStateChange) => {
-        if (change.current === 'attached') {
-          contributor.channel.off(listener);
-          resolve();
-          return;
-        }
-
-        if (change.current === 'failed') {
-          contributor.channel.off(listener);
-          this._lifecycle.setStatus({ status: RoomStatus.Failed, error: change.reason });
-
-          // Its ok to just set operation in progress = false and return here
-          // As every other channel is wound down.
-          this._endLifecycleOperation();
-          throw change.reason ?? new Ably.ErrorInfo('unknown error in _doRetry', ErrorCodes.RoomLifecycleError, 500);
-        }
-      };
-      contributor.channel.on(listener);
-    }).then(() => {
-      this._logger.debug('RoomLifecycleManager._doRetry(); feature reattached via listener, retrying attach');
-      return doAttachWithRetry();
+        this._logger.warn('discontinuity detected', {
```

Comment:
Looks like a few logs in this file don't follow the standard format
```suggestion
        this._logger.warn('_startMonitoringDiscontinuity(); discontinuity detected', {
```

Reply:
[6b7eb10](https://github.com/ably/ably-chat-js/pull/505/commits/6b7eb10edf6786264adf942557fc7e8a17d44d36)